### ISSUE TITLE
Using busybox-w32 as BAZEL_SH

### ISSUE THREAD

**Issue by Xjs** [bazelbuild/bazel#7274]
I tried using https://github.com/rmyorston/busybox-w32 as a replacement for bash as BAZEL_SH. After uncovering some problems with Windows line feed handling, which fortunately were really quickly addressed by busybox-w32 maintainers (cf. https://github.com/rmyorston/busybox-w32/issues/138) I thought it would work. However, I keep getting »Permission denied« errors when using `git_repository` in my WORKSPACE.

I have set up a repository that demonstrates the issue: https://github.com/Xjs/bazel-busybox-git

Steps to reproduce:

* Check out bazel-busybox-git
* Set `BAZEL_SH` to the path of `busybox-0aad680c.exe` contained within
* Execute `bazel build main`

Relevant line of the failure output:

```rm: can't remove 'C:/tmp/_bazel_jannis.schnitzer/fji6e52k/external/com_github_Xjs_dummy_dependency': Permission denied (2x)
```

(full output see https://github.com/Xjs/bazel-busybox-git)

What could cause this? How could I investigate further? (Debugging git_repository is hard because Bazel checks the tools/*.bzl files for integrity)

**Comment by ale5000-git** [comment]
I don't really know exactly your problem but usually when a "deletion" command fails on Windows but works on Linux it is because on Linux you can delete an "in use" files while on Windows you cannot do it.

**Comment by rmyorston** [comment]
I think @ale5000-git is probably correct. I've had a look at git.bzl and suspect the directory that can't be removed is the current directory of the process.

MSYS2 is based on Cygwin. It uses clever tricks so the following works as it would on Unix and the directory really does get deleted:
```
$ mkdir xxx
$ cd xxx
$ rm -rf ../xxx
```
busybox-w32 doesn't aspire to such feats and just reports 'Permission denied'. This is not something that's likely to change.

**Comment by ale5000-git** [comment]
In my scripts I put the deletion command of a folder after it is no longer used by the code and I use "cd" to switch to a different directory (not the one that must be deleted) before deleting it.

**Comment by Xjs** [comment]
Oh. Hang on. The working directory of the BAZEL_SH is `C:/tmp/_bazel_jannis.schnitzer/fji6e52k/external/com_github_Xjs_dummy_dependency`, and `cd C:/tmp/_bazel_jannis.schnitzer/fji6e52k/external` is already called in a sub-shell.

It's nice of Cygwin to do some tricks to make this work, but I think this should be fixed by Bazel as to not require this feature (deleting a directory while inside) anymore.

**Comment by Xjs** [comment]
Related: https://github.com/bazelbuild/bazel/issues/6339

In a nutshell: @laszlocsomor said in the thread there, `git_repository` shouldn't depend on a shell at all. While it does, it should also not try to delete the current working directory.

**Comment by rmyorston** [comment]
>The working directory of the BAZEL_SH is ```C:/tmp/_bazel_jannis.schnitzer/fji6e52k/externalcom_github_Xjs_dummy_dependency```

Right, that's what I suspected. I don't know enough about Bazel to figure out how that comes about but it does explain the observed behaviour.

**Comment by dslomov** [comment]
We should fix this as a part of general effort to not depend on bash for repository rules.

**Comment by laszlocsomor** [comment]
> We should fix this as a part of general effort to not depend on bash for repository rules.

That effort (i.e. not depending on Bash) is tracked in https://github.com/bazelbuild/bazel/issues/4319.

**Comment by dslomov** [comment]
Oh wow I missed that there is already a proposed fix (#7285) - very cool!

**Comment by Xjs** [comment]
@dslomov Seems not to work in all cases yet, though. I'll put some more work in.
